File: ios/Classes/WGEventChannel.swift
//
//  WGEventChannel.swift
//  wireguard_dart
//
//  Created by Daniil on 05.07.2023.
//

import Foundation
import Flutter

class WGEventChannel : NSObject, FlutterStreamHandler {
    static var sink: FlutterEventSink?
    
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        Self.sink = events
        return nil
    }
    
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        Self.sink = nil
        return nil
    }
    
    static func sendState(state: String) {
        guard let sink = Self.sink else {return}
        
        sink(state)
    }
    
}

File: android/src/main/kotlin/network/mysterium/wireguard_dart/WGEventChannel.kt
package network.mysterium.wireguard_dart

import android.os.Handler
import io.flutter.plugin.common.EventChannel
import java.util.*

class WGEventChannel : EventChannel.StreamHandler {
    var sink: EventChannel.EventSink? = null

    fun send(state: String) {
        when (state) {
            "UP" -> sink?.success("connected")
            "DOWN" -> sink?.success("disconnected")
            "TOGGLE" -> sink?.success("connecting")
            else -> sink?.success("none")
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        sink = events
    }

    override fun onCancel(arguments: Any?) {
        sink = null
    }
}